File: backend/src/main/java/com/skewwhiffy/auraltester/notation/model/note/NoteSequence.java
package com.skewwhiffy.auraltester.notation.model.note;

import java.util.Arrays;
import java.util.List;

public interface NoteSequence {
    List<AbsoluteNote> getNotes();
    /*
interface NoteSequence {
  val notes: List<AbsoluteNote>

  companion object {
    val empty: NoteSequence = object : NoteSequence {
      override val notes: List<AbsoluteNote> = listOf()
    }

  }
}
     */
    static NoteSequence of(AbsoluteNote... notes) {
        return () -> Arrays.asList(notes);
    }
}

File: backend/src/main/kotlin/com/skewwhiffy/auraltester/notation/model/note/NoteSequence.kt
package com.skewwhiffy.auraltester.notation.model.note;

interface NoteSequence {
    companion object {
        private data class NoteSequenceImpl(override val notes: List<AbsoluteNote>) : NoteSequence
        fun of(vararg notes: AbsoluteNote): NoteSequence {
            return NoteSequenceImpl(notes.toList())
        }
    }

    val notes: List<AbsoluteNote>
}
